### DISCUSSION TITLE
How to override ImageMagick's policy.xml on Streamlit

### DISCUSSION THREAD

**Discussion by barteshwar** [ImageMagick/ImageMagick#6467] [Help]
### ImageMagick version

7

### Operating system, version and so on

Streamlit

### Description

Hi,

I've been trying to build a python app using moviepy. The moviepy library seems to use ImageMagick for its TextClip function. Now using this function doesn't seem to be compatible with ImageMagick's latest versions, and throws this errror:

@/tmp/tmp6ca_nccx.txt' @ error/property.c/InterpretImageProperties/3708. convert-im6.q16: no images defined PNG32:/tmp/tmplzoyr257.png’ @ error/convert.c/ConvertImageCommand/3229.
.This error can be due to the fact that ImageMagick is not installed on your computer, or (for Windows users) that you didn’t specify the path to the ImageMagick binary in file conf.py, or that the path you specified is incorrect

On Google Colab, I'd fixed this error by manually updating the policy.xml file by updating read|write in this line:
 <policy domain="path" rights="read|write" pattern="@*"/>

On Windows, I was able to bypass this error by selecting add legacy configurations during ImageMagick's installation.

Now I'm trying to deploy the app on Streamlit. There seems to be no way to access the policy.xml file on Streamlit. Does anyone know how I can change policy.xml so that my code runs on Streamlit?

Any pointers will help! Thanks!

**Comment by fmw42** [comment]
I suspect moviepy is looking for Imagemagick 6 and you are using Imagemagick 7.  I think you need to contact the moviepy people and ask for an update using Imagemagick 7.  On Windows, enabling the "legacy components" installs Imagemagick 6. That is why that works.

Sorry, I know nothing about Streamlit.

**Comment by snibgo** [comment]
The MoviePy TextClip documentation at https://moviepy-tburrows13.readthedocs.io/en/improve-docs/ref/VideoClip/TextClip.html says:

> Creates an ImageClip originating from a script-generated text image. Requires ImageMagick.

Your error messages ...

> @/tmp/tmp6ca_nccx.txt' @ error/property.c/InterpretImageProperties/3708.
convert-im6.q16: no images defined PNG32:/tmp/tmplzoyr257.png’
@ error/convert.c/ConvertImageCommand/3229.

... seem to come from ImageMagick. So IM is being executed. But ...

> This error can be due to the fact that ImageMagick is not installed on your computer, or (for Windows users) that you didn’t specify the path to the ImageMagick binary in file conf.py, or that the path you specified is incorrect

That seems to be incorrect. IM has been found and executed.

What IM command is MoviePy issueing?

If IM can't access a required font, it issues a warning, not an error.

What is the text output from "`convert -version`"?

Does "`convert -list font`" include your expected fonts?

Have you *installed* your fonts in Windows, or merely copied them to the Windows directory?

**Comment by barteshwar** [comment]
Hi,

Thanks for the replies. It doesn't seem to be about the font. I removed the
font specification, and I am still getting the same error.

It appears that what's causing this error is ImageMagick's policy.xml. I
had got the same errors on both Google Colab and my local machine, and I
had fixed them by updating ImageMagick's policy.xml by changing this line:

 <policy domain="path" rights="none" pattern="@*"/>

to this line

 <policy domain="path" rights="read | write" pattern="@*"/>

The problem with streamlit is that I'm unable to access the policy.xml
file, so I'm not able to make the change. Would you know how I could access
the policy.xml if my app is deployed on Streamlit, or if there are any
other workarounds? I've also tried forcing Streamlit to install the older
version of ImageMagick (version 6), but I'm still getting the same error
message.

Thanks for your help.

**Comment by snibgo** [comment]
> The problem with streamlit is that I'm unable to access the policy.xml file, ...

Is that because you don't have authority to change files in that directory? Why not? Perhaps because that directory is used by other users, so there is a security problem if users can edit those files?

The only workaround is to install ImageMagick in a private directory, and to ensure you use that version.

**Comment by barteshwar** [comment]
I'm a relative beginner, and I'm trying to deploy an app on Streamlit for
the first time. As far as I can tell, it doesn't allow you to edit
installed packages. I don't even seem to have access to a terminal. Would
you happen to know how I could edit the policy.xml file?

Also, how do I install ImageMagick in a private directory? Right now, I'm
installing ImageMagick by including the text "imagemagick" in a file called
packages.txt, and Streamlit automatically uses apt-get (I think) to install
it from there.

Thanks again for your help!

**Comment by snibgo** [comment]
Sorry, I know nothing about Streamlit.

I don't know how to install ImageMagick without access to a terminal.

[EDIT: I should qualify that. Windows binary distributions come as ".msi" files. If you click on the msi icon, a window opens that guides the user through the installation process. So that doesn't need access to a terminal. ]

**Comment by barteshwar** [comment]
Oh okay. Thanks anyway!
